graphjet-core/src/main/java/com/twitter/graphjet/algorithms/PageRank.java

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright 2016 Twitter. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.twitter.graphjet.algorithms;
+
+import com.google.common.util.concurrent.AtomicDouble;
+import com.twitter.graphjet.bipartite.api.EdgeIterator;
+import com.twitter.graphjet.directed.api.OutIndexedDirectedGraph;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import it.unimi.dsi.fastutil.longs.LongIterator;
+import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
+
+/**
+ * Implementation of PageRank. This implementation initializes an array of doubles to hold PageRank values, where the
+ * node id is used as the index into the array. Therefore, the algorithm needs to know the max node id as well as the
+ * set of nodes, in the case where the node ids are not assigned sequentially.
+ */
+public class PageRank {
+  final private OutIndexedDirectedGraph graph;
+  final private LongOpenHashSet nodes;
+  final private long maxNodeId;
+  final private double dampingFactor;
+  final private int nodeCount;
+  final private int maxIterations;
+  final private double tolerance;
+
+  final private AtomicDouble normL1 = new AtomicDouble(Double.MAX_VALUE);
+  private double[] prVector = null;
+
+  /**
+   * Constructs this object for running PageRank over a directed graph.
+   *
+   * @param graph          the directed graph
+   * @param nodes          nodes in the graph
+   * @param maxNodesId     maximum node id
+   * @param dampingFactor  damping factor
+   * @param maxIterations  maximum number of iterations to run
+   * @param tolerance      L1 norm threshold for convergence
+   */
+  public PageRank(OutIndexedDirectedGraph graph, LongOpenHashSet nodes, long maxNodesId, double dampingFactor,
```

Comment:
That is correct. I've already described this in the top-level documentation of the class, but probably worth elaborating.